### ISSUE TITLE
Mocking WCF Service Class Interface

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ameniameni/moq#8]
```
What steps will reproduce the problem?
1. Mock<IServiceClassInterface> mock = new Mock<IServiceClassInterface>();
2. ServiceHost host = new ServiceHost(mock.Object);
3. host.Open();

What is the expected output? What do you see instead?

I'd like it to open the host but instead it throws the following error..

Test method MockTest.TestMockIServiceClassInterface threw exception:  
System.InvalidOperationException: The service class of type 
IServiceClassInterfaceProxya3beb8cf8f8349a39ee8219439225508 both defines a 
ServiceContract and inherits a ServiceContract from type 
IServiceClassInterface. Contract inheritance can only be used among 
interface types.  If a class is marked with ServiceContractAttribute, it 
must be the only type in the hierarchy with ServiceContractAttribute.  
Consider moving the ServiceContractAttribute on type 
IServiceClassInterface to a separate interface that type 
IServiceClassInterface implements..


What version of the product are you using? On what operating system?

1.5.1 on Vista x64

Please provide any additional information below.

This doesn't work with RhinoMock 3.3 either but it did work with a 
previous version - this is 99% of all mocking for my project now so it's 
pretty important and should be to all others building SOAs.

** I did try mocking the type rather than the interface also which seems 
to open the host but then not intercept the calls.  Am I supposed to pass 
in the mock or mock.Object for the singletonInstance of the ServiceHost 
also??  I'd much prefer to be able to mock the interface as that doesn't 
require me to either have a reference to the actual service class or to 
create an empty implementation just for the purpose of mocking ...

I suspect this is all due to Castle copying all attributes from the 
interface to the mock.Object implementation .. it would be awesome if you 
could either give me a way to ignore specific ones or at least ignore this 
one?? or check the attribute properties to see if multiple are allowed 
then automatically ignore it if not.. so many options.

Thanks! Boom
```

Original issue reported on code.google.com by `freef...@googlemail.com` on 26 Feb 2008 at 3:49

**Comment by GoogleCodeExporter** [comment]
```
Thanks a lot for an excellent bug report!

If this doesn't work on Rhino, rest assured it's because of Castle 
DynamicProxy2.
We'll look into customizing its behavior to allow for this. Your suggestion on
looking at the "AllowMultiple" property on attributes seems like a very good 
approach
indeed.

We'll follow up when we have more data and an approach to solve this.
```

Original comment by `kzu.net` on 27 Feb 2008 at 2:40
- Added labels: **Milestone-Release2.0**, **Priority-High**
- Removed labels: **Priority-Medium**

**Comment by GoogleCodeExporter** [comment]
```
That's a big high five kzu, thanks

After a bit more investigation I've managed a fix all by myself - unfortunately 
it 
is in the castle project but maybe you can talk to them??  anyway the fix is 
below, 
maybe you can sneak it into the next build...

class: Castle.DynamicProxy.Generators.BaseProxyGenerator

private bool ShouldSkipAttributeReplication(Attribute attribute)
{
    if (SpecialCaseAttributThatShouldNotBeReplicated(attribute))
        return true;

    object[] attrs = attribute.GetType()
        .GetCustomAttributes(typeof(AttributeUsageAttribute), true);

    if (attrs.Length != 0)
    {
        AttributeUsageAttribute usage = (AttributeUsageAttribute) attrs[0];

        return usage.Inherited || !usage.AllowMultiple;
    }

    return true;
}

The fix is this bit || !usage.AllowMultiple

Cheers
```

Original comment by `freef...@googlemail.com` on 27 Feb 2008 at 12:06

**Comment by GoogleCodeExporter** [comment]
```
It may be easier if you report this directly to Rhino, as its developer 
(Ayende) is
also a contributor to Castle... he has more chances of having this fixed than 
me :).
But I'll try.
```

Original comment by `kzu.net` on 28 Feb 2008 at 10:56

**Comment by GoogleCodeExporter** [comment]
Original comment by `kzu.net` on 7 Mar 2008 at 5:57
- Removed labels: **Milestone-Release2.0**

**Comment by GoogleCodeExporter** [comment]
Original comment by `kzu.net` on 10 Mar 2008 at 4:43
- Changed state: **Fixed**
- Added labels: **Milestone-Release2.0**

**Comment by GoogleCodeExporter** [comment]
```
cool beans!
```

Original comment by `freef...@googlemail.com` on 11 Mar 2008 at 8:57

**Comment by GoogleCodeExporter** [comment]
```
Let us know if everything works as expected!
```

Original comment by `kzu.net` on 11 Mar 2008 at 1:21

**Comment by GoogleCodeExporter** [comment]
```
good as gold....it's in production

thanks kzu!
```

Original comment by `freef...@googlemail.com` on 11 Mar 2008 at 5:21

**Comment by GoogleCodeExporter** [comment]
```
It seems as this issue is not fixed yet in the current version of Moq (3.0 
beta) or 
in Rhino Mocks for that matter. Can anyone provide some status information on 
it and 
maybe reopen the issue 
```

Original comment by `Bakker.F...@gmail.com` on 16 Feb 2009 at 3:16

**Comment by GoogleCodeExporter** [comment]
```
i agree .. not fixed yet. had to re-apply the fix in the latest version
```

Original comment by `freef...@googlemail.com` on 16 Feb 2009 at 3:34
